commands/initcmd.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package commands
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile"
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile/kptfileutil"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/cli-utils/pkg/config"
+)
+
+const defaultInventoryName = "inventory"
+
+// KptInitOptions encapsulates fields for kpt init command. This init command
+// fills in inventory values in the Kptfile.
+type KptInitOptions struct {
+	dir                string // Directory with Kptfile
+	force              bool   // Set inventory values even if already set in Kptfile
+	name               string // Inventory object name
+	namespace          string // Inventory object namespace
+	inventoryID        string // Inventory object unique identifier label
+	defaultInitOptions *config.InitOptions
+}
+
+// NewKptInitOptions returns a pointer to an initial KptInitOptions structure.
+func NewKptInitOptions(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *KptInitOptions {
+	io := config.NewInitOptions(f, ioStreams)
+	return &KptInitOptions{
+		defaultInitOptions: io,
+	}
+}
+
+// Complete fills in fields for KptInitOptions based on the passed "args".
+func (io *KptInitOptions) Complete(args []string) error {
+	if err := io.defaultInitOptions.Complete(args); err != nil {
+		return err
+	}
+	io.dir = io.defaultInitOptions.Dir
+	// Add random suffix to default inventory name to avoid collisions
+	if io.name == defaultInventoryName {
+		randomSuffix := common.RandomStr(time.Now().UTC().UnixNano())
+		io.name = fmt.Sprintf("%s-%s", defaultInventoryName, randomSuffix)
+	}
+	io.namespace = strings.TrimSpace(io.defaultInitOptions.Namespace)
+	io.inventoryID = strings.TrimSpace(io.defaultInitOptions.InventoryID)
```

Comment:
So the id as we want to use for the Kptfile version should be based on the actual values, and not just a UID like we do with the inventory template solution.

Reply:
Please explain this. I do not understand your comment.

Reply:
Based on the discussions and the design, we want the generated id to be a combination of the hash of the name+namespace and the timestamp when this command is run. That will allow us to use the id as a checksum and detect situations when either the name or the namespace has been changed.

Reply:
Initial implementation of `generateID()` function to create the `namespace/name`+`unixtimestamp` identifier (plus unit tests).